### PR TITLE
feat: Add notification token callback to use en addMessage Method

### PR DESCRIPTION
**Pull request by gabrielmatau79**
Add new callback into Message Pickup Repository Client to obtain FCM device token from Didcomm Mediator.

### CODE REVIEW

File: packages/client/src/MessagePickupRepositoryClient.ts [hunk 1]
```diff
@@ -94,6 +104,27 @@ export class MessagePickupRepositoryClient implements MessagePickupRepository {
...
+   * const client = new MessagePickupRepositoryClient({ url: 'wss://example.com' });
+   *
+   * const getDeviceToken = async (connectionId: string) => {
+   *   const connectionRecord = await agent.connections.findById(connectionId);
+   *   return connectionRecord?.getTag('device_token') as string | undefined;
+   * };
+   *
+   * client.notificationToken(getDeviceToken);
+   */
+  notificationToken(callback: (connectionId: string) => Promise<string | undefined>): void {
```
Comment (genaris) [inline packages/client/src/MessagePickupRepositoryClient.ts]: I think we can make it more generic and call it `connectionInfoCallback`.

```suggestion
  setConnectionInfoCallback(callback: (connectionId: string) => Promise<ConnectionInfo | undefined>): void {
     this.connectionInfoCallback = callback
  }
```

Where:

```
export type ConnectionInfo = {
   fcmNotificationToken?: string
   maxReceiveBytes?: number
}
```

**Review by genaris — CHANGES_REQUESTED** [review]
I think it is great, but just wondering on some updates to the prototype of the callback in order to make it useful for other information, such as the `maxReceiveBytes` and maybe other fields we may add on the future, without introducing too much breaking changes.

File: packages/client/src/MessagePickupRepositoryClient.ts (hunk 1 above)
Comment (gabrielmatau79) [reply packages/client/src/MessagePickupRepositoryClient.ts]: Solved! 

In this update, I implemented a more generic callback function setConnectionInfo to allow dynamic retrieval of connection-specific information, such as the fcmNotificationToken and maxReceiveBytes. This callback replaces the previously token-specific callback, making it more versatile and adaptable for various connection-related data needs.

Key modifications:

- Introduced ConnectionInfo interface, which includes properties like fcmNotificationToken and maxReceiveBytes.

- Added setConnectionInfo method in MessagePickupRepositoryClient, allowing the client to set a callback that fetches ConnectionInfo based on connectionId.

- Updated addMessage and takeFromQueue methods to use the setConnectionInfoCallback for obtaining necessary connection data dynamically.

- Remove maxReceiveBytes definition of the constructor class MessagePickupRepositoryClient

File: packages/client/src/MessagePickupRepositoryClient.ts [hunk 2]
```diff
@@ -21,12 +23,11 @@ export class MessagePickupRepositoryClient implements MessagePickupRepository {
   private client?: Client
   private readonly logger = log
   private messagesReceivedCallback: ((data: MessagesReceivedCallbackParams) => void) | null = null
+  private setConnectionInfoCallback?: (connectionId: string) => Promise<ConnectionInfo | undefined>
```
Comment (genaris) [inline packages/client/src/MessagePickupRepositoryClient.ts]:
```suggestion
  private connectionInfoCallback?: (connectionId: string) => Promise<ConnectionInfo | undefined>
```

File: packages/client/src/MessagePickupRepositoryClient.ts [hunk 3]
```diff
@@ -116,9 +151,15 @@ export class MessagePickupRepositoryClient implements MessagePickupRepository {
     try {
       const client = this.checkClient()
 
+      const connectionInfoCallback = this.setConnectionInfoCallback
+        ? await this.setConnectionInfoCallback(params.connectionId)
+        : undefined
+
+      const maxReceiveBytes = connectionInfoCallback?.maxReceiveBytes
```
Comment (genaris) [inline packages/client/src/MessagePickupRepositoryClient.ts]:
```suggestion
      const connectionInfo = this.connectionInfoCallback
        ? await this.connectionInfoCallback(params.connectionId)
        : undefined

      const maxReceiveBytes = connectionInfo?.maxReceiveBytes
```

File: packages/client/src/MessagePickupRepositoryClient.ts [hunk 4]
```diff
@@ -164,33 +205,43 @@ export class MessagePickupRepositoryClient implements MessagePickupRepository {
...
       const client = this.checkClient()
-      // Call the RPC method and store the result as 'unknown' type initially
+
+      // Retrieve connection information using the callback, if set
+      const connectionInfoCallback = this.setConnectionInfoCallback
+        ? await this.setConnectionInfoCallback(params.connectionId)
+        : undefined
+
+      // Set the token and max bytes from the connection info, if available
+      params.token = connectionInfoCallback?.fcmNotificationToken
```
Comment (genaris) [inline packages/client/src/MessagePickupRepositoryClient.ts]:
```suggestion
      const connectionInfo = this.connectionInfoCallback
        ? await this.connectionInfoCallback(params.connectionId)
        : undefined

      // Set the token and max bytes from the connection info, if available
      params.token = connectionInfo?.fcmNotificationToken
```

**Review by genaris — CHANGES_REQUESTED** [review]

**Review by genaris — APPROVED** [review]

**Pull request merged**